CMakeLists.txt

```
@@ -26,82 +24,165 @@ if(HIOP_USE_MPI)
     find_package(MPI REQUIRED)
     set(CMAKE_CXX_COMPILER ${MPI_CXX_COMPILER})
   endif(NOT DEFINED MPI_CXX_COMPILER)
-  file(APPEND "src/Interface/hiop_defs.hpp" "#define HIOP_USE_MPI\n")
-else()
-  file(APPEND "src/Interface/hiop_defs.hpp" "#ifdef HIOP_USE_MPI\n#undef HIOP_USE_MPI\n#endif\n")
 endif(HIOP_USE_MPI)
 
+# The binary dir is already a global include directory
+configure_file(
+  "${CMAKE_SOURCE_DIR}/src/Interface/hiop_defs.hpp.in"
+  "${CMAKE_BINARY_DIR}/hiop_defs.hpp")
+
+add_library(hiop_magma INTERFACE)
+add_library(hiop_math INTERFACE)
+add_library(hiop_cuda INTERFACE)
+add_library(hiop_options INTERFACE)
+add_library(hiop_warnings INTERFACE)
+
+target_compile_features(hiop_options
+    INTERFACE
+        cxx_std_11
+        cxx_alignas
+        cxx_alignof
+        cxx_attributes
+        cxx_auto_type
+        cxx_constexpr
+        cxx_defaulted_functions
+        cxx_deleted_functions
+        cxx_final
+        cxx_lambdas
+        cxx_noexcept
+        cxx_override
+        cxx_range_for
+        cxx_rvalue_references
+        cxx_static_assert
+        cxx_strong_enums
+        cxx_trailing_return_types
+        cxx_unicode_literals
+        cxx_user_literals
+        cxx_variadic_macros
+)
+
+target_compile_options(hiop_warnings
+    INTERFACE
+        -Wall
+        -Wextra
+        -Wshadow            # warn the user if a variable
+                            # declaration shadows one from a
+                            # parent context
+        -Wnull-dereference
+        -Wdouble-promotion  # Warn on implicit conversion from
+                            # float to double
+)
+
+if("${CMAKE_CXX_COMPILER_ID}" STREQUAL "GNU")
+    message("Using extra GNU-specific compiler warnings...")
+    target_compile_options(hiop_warnings
+        INTERFACE
+            -Wmisleading-indentation    # Warn on indented blocks
+                                        # that are not really blocks
+            -Wduplicated-cond           # Warn on if/else chain with
+                                        # duplicated conditions
+            -Wduplicated-branches       # Warn on if/else chains with
+                                        # duplicated code
+            -Wuseless-cast              # Warn when casting to the same type
+    )
+endif()
+
+if(HIOP_DEVELOPER_MODE)
+    target_link_libraries(hiop_math INTERFACE hiop_options hiop_warnings)
+endif()
+
 if(HIOP_USE_GPU)
+  include(CheckLanguage)
+  enable_language(CUDA)
+  check_language(CUDA)
+  if(NOT DEFINED CMAKE_CUDA_STANDARD)
+    set(CMAKE_CUDA_STANDARD 11)
+    set(CMAKE_CUDA_STANDARD_REQUIRED ON)
+  endif()
+
```

I think you still need:
```suggestion
find_package(CUDA REQUIRED) # CUDAToolkit as of 3.17
```
With this you get list of libraries that you can add to `hiop_cuda` interface.